onnxruntime/core/providers/nnapi/nnapi_builtin/model.h

```diff
@@ -123,7 +123,7 @@ class Model {
   std::unordered_map<std::string, android::nn::wrapper::OperandType> operand_types_;
   std::unordered_set<std::string> scalar_outputs_;
 
-  Shaper shaper_;
+  std::optional<Shaper> shaper_;
```

Comment:
actually, how about keeping the data member as `Shaper shaper_;` and updating Model to have a constructor like this?
`Model::Model(const GraphViewer& graph_viewer) : ..., shaper_{graph_viewer} {}`

then in `ModelBuilder::Prepare()`, create the model with `nnapi_model_ = std::make_unique<Model>(graph_viewer_);` and use nnapi_model_'s Shaper instance after that. maybe ModelBuilder doesn't even need its own Shaper

Reply:
Agree. 

The ownership of the Shaper needs to be clarified so this doesn't need to be `optional`. I agree with Ed that the Model should own it given it needs it for the longest time. I don't see why we need to create the model inside ModelBuilder::Prepare though given it's a call to the default ctor.

Based on that I would consider creating the model directly in NnapiExecutionProvider::Compile with the Model ctor taking a GraphViewer and creating Model::shaper_. Add a public getter for the Shaper to Model. ModelBuilder can use the Shaper instance from Model. Pass a Model& to the ModelBuilder ctor or ModelBuilder::Prepare. I'd prefer the former if we create the model prior to the ModelBuilder in Compile.